lib/gringotts/gateways/pin_pay.ex

```diff
@@ -0,0 +1,260 @@
+  ```
+  iex> card = %CreditCard{first_name: "Jo",
+                          last_name: "Doe",
+                          number: "4200000000000000",
+                          year: 2099, month: 12,
+                          verification_code: "123", brand: "VISA"}
+  iex> money = Money.new(20, :USD)
+  iex> {:ok, auth_result} = Gringotts.authorize(Gringotts.Gateways.PinPayments, amount, card, opts)
+  ```
+  """
+
+  @spec authorize(Money.t(), CreditCard.t() | String.t(), keyword) :: {:ok | :error, Response}
+  def authorize(amount, %CreditCard{} = card, opts) do
+    {currency, value, _} = Money.to_integer(amount)
+
+    card_token =
+      commit(:post, "cards", card_for_token(card, opts) ++ Keyword.delete(opts, :address))
+      |> extract_card_token
+
+    params =
+      [
+        amount: value,
+        capture: false,
+        card_token: card_token,
+        currency: currency
+      ] ++ Keyword.delete(opts, :address)
+
+    commit(:post, "charges", params)
+  end
+
+  def authorize(amount, card_token, opts) when is_binary(card_token) do
+    {currency, value, _} = Money.to_integer(amount)
+
+    params =
+      [
+        amount: value,
+        capture: false,
+        currency: currency,
+        card_token: card_token
+      ] ++ Keyword.delete(opts, :address)
+
+    commit(:post, "charges", params)
+  end
+
+  
+  ###############################################################################
+  #                                PRIVATE METHODS                              #
+  ###############################################################################
+
+  # Makes the request to PinPay's network.
+  # For consistency with other gateway implementations, make your (final)
+  # network request in here, and parse it using another private method called
+  # `respond`.
+
+  defp card_for_token(card, opts) do
+    [
+      number: card.number,
+      name: CreditCard.full_name(card),
+      expiry_month: card.month |> Integer.to_string() |> String.pad_leading(2, "0"),
+      expiry_year: card.year |> Integer.to_string(),
+      cvc: card.verification_code,
+      address_line1: opts[:Address][:street1],
+      address_city: opts[:Address][:city],
+      address_country: opts[:Address][:country],
+      address_line2: opts[:Address][:street2],
+      address_postcode: opts[:Address][:postal_code],
+      address_state: opts[:Address][:region]
+
+    ]
+  end
+
+  @spec commit(atom, String.t(), keyword) :: {:ok | :error, Response}
+  defp commit(:post, endpoint, param) do
+    auth_token = encoded_credentials(param[:config].apiKey, "")
+
+    headers = [
+      {"Content-Type", "application/x-www-form-urlencoded"},
+      {"Authorization", auth_token}
+    ]
+
+    url = @test_url <> "#{endpoint}"
+    param = Keyword.delete(param, :config)
+
+    url
+    |> HTTPoison.post({:form, param}, headers)
+    |> respond
+  end
+
+  defp commit_short(method, url, opts) do
+    auth_token = encoded_credentials(opts[:config].apiKey, "")
```

Comment:
Your moduledoc says the API key must be provided under the key `username` (see L52), but your code is using some undocumented `:apiKey`?